File: 2-rotor-system-simulator/src/rotor_sim.py
from time import time
from typing import Callable


class RotorSim:

    def __init__(
        self,
        after_turn_consumer: Callable,
    ) -> None:
        self.after_turn_consumer = after_turn_consumer

    def f3(self, rotors: list, rotor_size: int, rotors_max_num: int):
        """Combine the add and append_rotor while loops into one
        # 1
        """

        start_time = time()

        while len(rotors) <= rotors_max_num:

            self.after_turn_consumer(rotors)

            # Turn base rotor
            rotors[0] += 1
            rotors[0] %= rotor_size

            # Turn non-base rotors
            i = 0
            add_rotor_flag = False
            while i < len(rotors):
                if rotors[i] != 0:
                    add_rotor_flag = False
                    break

                if (len(rotors) > 1) and (i < len(rotors) - 1):
                    rotors[i + 1] += 1
                    rotors[i + 1] %= rotor_size

                if i == len(rotors) - 1:
                    if rotors[i] == 0:
                        add_rotor_flag = True

                i += 1

            # Add rotor
            if add_rotor_flag:
                rotors.append(0)

        print(f"PYTHON ROTOR SIMULATOR")
        print(f"rotors: {rotors}")
        print(f"time to completion: {time()-start_time}")

File: 2-rotor-system-simulator/src/rotor_sim.c
#include <stdbool.h>
#include <stdio.h>

void rotor_sim(
    int *p_starter,
    int len_starter,
    int rotor_size,
    int rotors_max_num,
    void (*p_turn_func)(int *, int, char),
    void (*p_append_func)(int *, int, char),
    void (*p_completion_func)(int *, int, char))
{
    // Initialize and populate ROTORS with the starter values if present
    int ROTORS[rotors_max_num];
    ROTORS[0] = 0;
    int LEN_ROTORS = 1;

    if (len_starter > 0)
    {
        LEN_ROTORS = 0;
        for (int i = 0; i < len_starter; i++)
        {
            ROTORS[i] = *(p_starter + i);
            LEN_ROTORS++;
        }
    }

    // Where the magic happens
    while (LEN_ROTORS <= rotors_max_num)
    {
        (*p_turn_func)(ROTORS, LEN_ROTORS, '\r');

        // Turn base rotor
        ROTORS[0]++;
        ROTORS[0] %= rotor_size;

        // Turn other rotors up the chain
        bool add_rotor_flag = false;
        for (int i = 0; i < LEN_ROTORS; i++)
        {
            if (ROTORS[i] != 0)
            {
                add_rotor_flag = false;
                break;
            }

            if (LEN_ROTORS > 1 && (i < LEN_ROTORS - 1))
            {
                ROTORS[i + 1]++;
                ROTORS[i + 1] %= rotor_size;
            }

            if (i == LEN_ROTORS - 1)
                if (ROTORS[i] == 0)
                    add_rotor_flag = true;
        }

        // Append rotor
        if (add_rotor_flag)
        {
            ROTORS[LEN_ROTORS] = 0;
            LEN_ROTORS++;
            (*p_append_func)(ROTORS, LEN_ROTORS, '\n');
        }
    }

    (*p_completion_func)(ROTORS, LEN_ROTORS, '\n');
}

void print_breaker_array(
    int *p_breaker_array,
    int len_breaker_array,
    char next_line)
{
}

void print_after_completion(
    int *p_breaker_array,
    int len_breaker_array,
    char next_line)
{
    printf("C ROTOR SIMULATOR \n");
    printf("rotors: ");
    for (int i = 0; i < len_breaker_array; i++)
    {
        printf("%d ", *(p_breaker_array + i));
        if (i == len_breaker_array - 1)
            printf("%c", next_line);
    }
}

int main()
{
    int starter[3] = {0};
    int starter_len = sizeof(starter) / sizeof(starter[0]);

    rotor_sim(
        starter,
        starter_len,
        10,
        7,
        &print_breaker_array,
        &print_breaker_array,
        &print_after_completion);
    return 0;
}
